propolis/src/block/mod.rs

```diff
@@ -46,9 +49,31 @@ pub type CompleteFn =
 
 /// Block device operation request
 pub struct Request {
+    /// The type of operation requested by the block device
     op: Operation,
+
+    /// A list of regions of guest memory to read/write into as part of the I/O request
     regions: Vec<GuestRegion>,
+
+    /// Block devuce specific completion function for this I/O request
     donef: Option<Box<CompleteFn>>,
+
+    /// A shared count of outstanding I/O requests for the block backend handling this request
+    ///
+    /// The `Request::new_*` methods explicitly initialize these but it is the backend
+    /// Notifier that will actually replace it with the correct reference.
+    /// See [`Request::attach_outstanding_notifier`].
+    outstanding_reqs: Arc<AtomicU64>,
+
+    /// A shared notifier to indicate to the block backend handling this request once
+    /// all outstanding requests have been completed (i.e. `outstanding_reqs` hits 0).
+    /// This is optional because we don't always want to be notified when we hit 0
+    /// outstanding requests but only if the device has been paused.
+    ///
+    /// The `Request::new_*` methods explicitly initialize these but it is the backend
+    /// Notifier that will actually replace it with the correct reference.
+    /// See [`Request::attach_outstanding_notifier`].
+    outstanding_notifier: Arc<Mutex<Option<Arc<Notify>>>>,
```

Comment:
Perhaps wrap the outstanding notification state into a purpose-built struct?  I think it would help simplify reasoning about it.

Then it could be something like
```rust
pub struct Request {
...
    outstanding_req_state: Option<Arc<OutstandingReqs>>
...
}

struct Outstanding Reqs {
    count: AtomicU64,
    notifier: Notify,
}
```

Reply:
Updated PR